seed-data-scripts/publication-metadata/script.js

```diff
@@ -0,0 +1,144 @@
+
+const createLogsFile = (logsArray) => {
+  fs.writeFile(LOGS_FILE, logsArray.join("\n"), (error) => {
+    if (error) {
+      console.log(error.message);
+    }
+  });
+};
+
+const getMetadata = async () => {
+  const fetchedDOIs = [];
+  const logs = [];
+
+  const addToLogs = (text) => {
+    const timestamp = new Date().toLocaleTimeString();
+    const message = `${timestamp}\t${text}`;
+    console.log(message);
+    logs.push(message);
+  };
+
+  try {
+    fs.readFile(FILE_NAME, "utf8", async (error, data) => {
+      if (error) throw error;
+
+      // extract doi urls as a Set from the entire seed data file
+      const doiUrls = [
+        ...new Set(getURLsFromText(data).filter((url) => validateDOI(url))),
+      ];
+
+      if (!doiUrls.length) {
+        // don't do anything
+        return;
+      }
+
+      const client = axios.create({
+        headers: {
+          Accept: "text/x-bibliography",
+          style: "harvard-cite-them-right",
+        },
+      });
+
+      axiosRetry(client, {
+        retries: 3,
+        shouldResetTimeout: true,
+        retryDelay: (retryCount, error) => {
+          if (error.response?.status === 429) {
+            // api requests limit reached https://support.datacite.org/docs/is-there-a-rate-limit-for-making-requests-against-the-datacite-apis
+            return 5 * 60 * 1000; // wait for 5 mins
+          }
+
+          return retryCount * 1000;
+        },
+        retryCondition: (error) => error.response?.status !== 404, // don't retry on 404
+      });
+
+      const nrOfChunks = Math.ceil(doiUrls.length / CHUNK_SIZE); // fetch urls in chunks
+      let cursor = 0;
+      for (let i = 0; i < nrOfChunks; i++) {
+        const currentChunk = doiUrls.slice(cursor, cursor + CHUNK_SIZE);
+
+        const promises = currentChunk.map((url) => {
+          addToLogs(`FETCHING => ${url}`);
+
+          return client.get(url).catch((error) => {
+            addToLogs(`ERROR => ${error.message} => ${url}`);
+            return error;
+          });
+        });
+
+        const results = await Promise.all(promises);
+
+        results.forEach((result, index) => {
+          if (!(result instanceof Error)) {
+            fetchedDOIs.push({ [currentChunk[index]]: result.data });
+          }
+        });
+
+        cursor += CHUNK_SIZE;
+        if (doiUrls.length - cursor > 0) {
+          addToLogs(`REMAINING ${doiUrls.length - cursor}`);
+        }
+      }
+
+      const parts = FILE_NAME.split(".");
+      const ext = parts[parts.length - 1];
+
+      fs.writeFile(
+        `${FILE_NAME.replace(`.${ext}`, "")}_references.json`,
+        JSON.stringify(fetchedDOIs),
+        { encoding: "utf8" },
+        async (error) => {
+          if (error) {
+            throw error;
+          }
+        }
+      );
+
+      createLogsFile(logs);
+    });
+  } catch (error) {
+    addToLogs(`ERROR => ${error.message}`);
+    createLogsFile(logs);
```

Comment:
could use a finally block here (then 136 not needed?)

Reply:
done